packages/headless/src/features/insight-user-actions/insight-user-actions-preprocessing.ts

```diff
@@ -0,0 +1,260 @@
+import {
+  UserAction,
+  UserActionsState,
+  UserActionTimeline,
+  UserActionType,
+  UserSession,
+} from './insight-user-actions-state';
+
+const MSEC_IN_SECOND = 1000;
+const SECONDS_IN_MINUTE = 60;
+const MAX_MINUTES_IN_SESSION = 30;
+const MAX_MSEC_IN_SESSION =
+  MAX_MINUTES_IN_SESSION * SECONDS_IN_MINUTE * MSEC_IN_SECOND;
+
+interface rawUserAction {
+  name: string;
+  value: string;
+  time: string;
+}
+
+export const preprocessActionsData = (
+  state: UserActionsState,
+  actions: Array<rawUserAction>
+): UserActionTimeline => {
+  const ticketCreationDate = state.ticketCreationDate;
+  const excludedCustomActions = state.excludedCustomActions;
+
+  // If there is no ticket creation date or no actions, we return an empty timeline.
+  if (!ticketCreationDate || actions.length === 0) {
+    return {
+      precedingSessions: [],
+      session: undefined,
+      followingSessions: [],
+      caseSessionFound: false,
+    };
+  }
+
+  const sortedActions: rawUserAction[] = sortActions(actions);
+
+  const mappedUserActions = mapUserActions(sortedActions);
```

Comment:
This should be a single function to go from O(2n) to O(n).
You can keep the filtering predicate and mapping/transforming function separate tho.

Reply:
I can merge the sort and mapping, but as I said in the other comment, the filtering needs to happen at the end. It is not so bad since we iterate the filtering over only the actions in the sessions in the timeline and not all the actions.